api-guidelines/compatibility.md

```diff
@@ -0,0 +1,161 @@
+    It may simplify implementation while providing freedom for further
+    evolution as compatible extensions.
+
+Not ignoring unknown input fields is a specific deviation from Postel’s
+Law (e.g. see also  
+[The Robustness Principle
+Reconsidered](https://cacm.acm.org/magazines/2011/8/114933-the-robustness-principle-reconsidered/fulltext))
+and a strong recommendation. Servers might want to take a different
+approach but should be aware of the following problems and be explicit
+in what is supported:
+
+  - Ignoring unknown input fields is actually not an option for PUT,
+    since it becomes asymmetric with subsequent GET response and HTTP is
+    clear about the PUT "replace" semantics and default roundtrip
+    expectations (see [RFC7231 Section
+    4.3.4](https://tools.ietf.org/html/rfc7231#section-4.3.4)). Note,
+    accepting (i.e. not ignoring) unknown input fields and returning it
+    in subsequent GET responses is a different situation and compliant
+    to PUT semantics.
+
+  - Certain client errors cannot be recognized by servers, e.g.
+    attribute name typing errors will be ignored without server error
+    feedback. The server cannot differentiate between the client
+    intentionally providing an additional field versus the client
+    sending a mistakenly named field, when the client’s actual intent
+    was to provide an optional input field.
+
+  - Future extensions of the input data structure might be in conflict
+    with already ignored fields and, hence, will not be compatible, i.e.
+    break clients that already use this field but with different type.
+
+In specific situations, where a (known) input field is not needed
+anymore, it either can stay in the API definition with "not used
+anymore" description or can be removed from the API definition as long
+as the server ignores this specific
+parameter.
+
+# SHOULD Avoid Versioning
+
+When changing your RESTful APIs, do so in a compatible way and avoid
+generating additional API versions. Multiple versions can significantly
+complicate understanding, testing, maintaining, evolving, operating and
+releasing our systems ([supplementary
+reading](http://martinfowler.com/articles/enterpriseREST.html)).
+
+If changing an API can’t be done in a compatible way, then proceed in
+one of these three ways:
+
+  - create a new resource (variant) in addition to the old resource
+    variant
+
+  - create a new service endpoint — i.e. a new application with a new
+    API (with a new domain name)
+
+  - create a new API version supported in parallel with the old API by
+    the same microservice
+
+As we discourage versioning by all means because of the manifold
+disadvantages, we strongly recommend to only use the first two
+approaches. Versioning should be a last resort.
+
+# MUST Use URI Versioning
+
+With URI versioning a (major) version number is included in the path,
+e.g. /v1/customers. If you don't put it in at the beginning then it's 
```

Comment:
Our current URL prefix is `/api/v1` etc. Would be nice to see here if we want to continue putting `/api` on things or not as well a being explicit that v1 goes at the start of the URL not just included

Reply:
I think this is up for discussion - personally, I only think you need /api at the beginning of the URL is if you are also serving out a website on the domain. Though I'd prefer a dedicated api. subdomain personally.